tools/eliminate-duplicate-functions.js

```diff
@@ -0,0 +1,524 @@
+    var equivalentFunctions = functionHashToFunctionName[functionHash];
+
+    // If we have more than one equivalent function, make sure
+    // that the bodies are the same from the hash values
+    if (equivalentFunctions.length > 1) {
+      var functionBody = getBodyForFunction(ast, equivalentFunctions[0]);
+
+      for (var functionIndex = 1; functionIndex < equivalentFunctions.length; ++functionIndex) {
+        var curFunctionBody = getBodyForFunction(ast, equivalentFunctions[functionIndex]);
+
+        if (curFunctionBody !== functionBody) {
+          printErr("ERROR!!! Function bodies for two equivalent functions differ!!! Candidates: "
+                  + equivalentFunctions[0] + ", " + equivalentFunctions[functionIndex]);
+          process.exit(1);
+        }
+      }
+    }
+  }
+}
+
+function eliminateDuplicateFuncs(ast) {
+  debugLog(ZONE_TOP_LEVEL, "eliminateDuplicateFuncs");
+
+  // Phase 1 - identify duplicate functions
+  var functionHashToFunctionName = identifyDuplicateFunctions(ast);
+
+  // Phase 2 - identify variables that conflict with function names
+  var variableNames = getVariableNames(ast);
+
+  // Phase 3 - generate the equivalent function hash
+  var equivalentFunctionHash = generateEquivalentFunctionHash(functionHashToFunctionName, variableNames);
+
+  // Phase 4 - for each set of equivalent functions, pick one and
+  // use it to replace the other equivalent functions.
+  replaceDuplicateFuncs(ast, equivalentFunctionHash);
+
+  return;
+}
+
+function find(filename) {
+  var prefixes = [nodePath.join(__dirname, '..', 'src'), process.cwd()];
+  for (var i = 0; i < prefixes.length; ++i) {
+    var combined = nodePath.join(prefixes[i], filename);
+    if (nodeFS.existsSync(combined)) {
+      return combined;
+    }
+  }
+  return filename;
+}
+
+function findAsmAst(ast) {
+  var asmNode = undefined;
+  traverse(ast, function(node, type) {
+    if (type === 'var') {
+
+      var vars = node[1];
+      for (var i = 0; i < vars.length; i++) {
+        var ident = vars[i][0];
+
+        if (ident === 'asm') {
+          asmNode = node;
+        }
+      }
+    }
+  });
+
+  return asmNode;
+}
+
+function printHashInfo(ast) {
+  debugLog(ZONE_TOP_LEVEL, "printHashInfo");
+
+  var infoHash = {};
+  infoHash['variable_names'] = getVariableNames(ast);
+  infoHash['fn_hash_to_fn_name'] = identifyDuplicateFunctions(ast);
+
+  print(JSON.stringify(infoHash));
+}
+
+read = function(filename) {
+  var absolute = find(filename);
+  return nodeFS['readFileSync'](absolute).toString();
+};
+
+// Main
+var arguments_ = process['argv'].slice(2);
+var noMinimizeWhitespace = false; // Eliminate whitespace by default
+
+// Expects a JS file from which we want the duplicate functions removed
+arguments_ = arguments_.filter(function(arg) {
+  if (arg === '--debug') {
+    debug = true;
+    debugFile = nodeFS.openSync(debugFileName, 'w');
+    return false;
+  } else if (arg === '--no-minimize-whitespace') {
```

Comment:
Yeah - we use this for running tests on DFE - all the input and output JS files do not have whitespace minimized for readability purposes.